src/debug_info.rs

```diff
@@ -1,30 +1,1060 @@
+        }
+    }
+}
+
+/// Any kind of debug info type
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DIType<'ctx> {
+    pub(crate) metadata_ref: LLVMMetadataRef,
+    _marker: PhantomData<&'ctx Context>,
+}
+
+impl<'ctx> DIType<'ctx> {
+    pub fn get_size_in_bits(&self) -> u64 {
+        use llvm_sys::debuginfo::LLVMDITypeGetSizeInBits;
+        unsafe { LLVMDITypeGetSizeInBits(self.metadata_ref) }
+    }
+
+    pub fn get_align_in_bits(&self) -> u32 {
+        use llvm_sys::debuginfo::LLVMDITypeGetAlignInBits;
+        unsafe { LLVMDITypeGetAlignInBits(self.metadata_ref) }
+    }
+
+    pub fn get_offset_in_bits(&self) -> u64 {
+        use llvm_sys::debuginfo::LLVMDITypeGetOffsetInBits;
+        unsafe { LLVMDITypeGetOffsetInBits(self.metadata_ref) }
+    }
+}
+
+/// A debug info typedef, created by `create_typedef` method of `DebugInfoBuilder`
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DIDerivedType<'ctx> {
+    pub(crate) metadata_ref: LLVMMetadataRef,
+    _marker: PhantomData<&'ctx Context>,
+}
+
+impl<'ctx> DIDerivedType<'ctx> {
+    pub fn as_type(&self) -> DIType<'ctx> {
+        DIType {
+            metadata_ref: self.metadata_ref,
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// A primitive debug info type created by `create_basic_type` method of `DebugInfoBuilder`
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DIBasicType<'ctx> {
+    pub(crate) metadata_ref: LLVMMetadataRef,
+    _marker: PhantomData<&'ctx Context>,
+}
+
+impl<'ctx> DIBasicType<'ctx> {
+    pub fn as_type(&self) -> DIType<'ctx> {
+        DIType {
+            metadata_ref: self.metadata_ref,
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// A debug info union type, created by `create_union_type` method of `DebugInfoBuilder`
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DICompositeType<'ctx> {
+    pub(crate) metadata_ref: LLVMMetadataRef,
+    _marker: PhantomData<&'ctx Context>,
+}
+
+impl<'ctx> DICompositeType<'ctx> {
+    pub fn as_type(&self) -> DIType<'ctx> {
+        DIType {
+            metadata_ref: self.metadata_ref,
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// Metadata representing the type of a function
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DISubroutineType<'ctx> {
+    pub(crate) metadata_ref: LLVMMetadataRef,
+    _marker: PhantomData<&'ctx Context>,
+}
+
+/// Lexical block scope for debug info
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DILexicalBlock<'ctx> {
+    pub(crate) metadata_ref: LLVMMetadataRef,
+    _marker: PhantomData<&'ctx Context>,
+}
+
+impl<'ctx> AsDIScope<'ctx> for DILexicalBlock<'ctx> {
+    fn as_debug_info_scope(self) -> DIScope<'ctx> {
+        DIScope {
+            metadata_ref: self.metadata_ref,
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// A debug location within the source code. Contains the following information:
+///
+/// - line, column
+/// - scope
+/// - is inlined
```

Comment:
Where it's been inlined into. The idea is that when code is initially parsed this is blank, then when the LLVM optimizer begins inlining it fills this in as it duplicates the LLVM IR frmo the callee into the caller. LLVM will use this to emit DWARF inlined subroutine debug info.